File: Dart/Structural/decorator.dart
/// Base interface
abstract interface class Notifier {
  void sendNotification();
}

/// Concrete implmenetation for SMS notifications
class SmsNotifier implements Notifier {
  @override
  void sendNotification() => print('Sending SMS notification!');
}

/// Concrete implementation for Facebook notifications
class FacebookNotifier implements Notifier {
  @override
  void sendNotification() => print('Sending Facebook notification!');
}

/// Concrete implementation for Slack notifications
class SlackNotifier implements Notifier {
  @override
  void sendNotification() => print('Sending Slack notification!');
}

/// Extension for SMS notifications
class SmsNotifierDecorator extends SmsNotifier {
  SmsNotifierDecorator({required this.notifier});

  final Notifier notifier;

  @override
  void sendNotification() {
    notifier.sendNotification();
    super.sendNotification();
  }
}

/// Extension for Facebook notifications
class FacebookNotifierDecorator extends FacebookNotifier {
  FacebookNotifierDecorator({required this.notifier});

  final Notifier notifier;

  @override
  void sendNotification() {
    notifier.sendNotification();
    super.sendNotification();
  }
}

/// Extension for Slack notifications
class SlackNotifierDecorator extends SlackNotifier {
  SlackNotifierDecorator({required this.notifier});

  final Notifier notifier;

  @override
  void sendNotification() {
    notifier.sendNotification();
    super.sendNotification();
  }
}

void main() {
  final smsNotifier = SmsNotifier();
  final facebookNotifierDecorator =
      FacebookNotifierDecorator(notifier: smsNotifier);
  final slackNotifierDecorator =
      SlackNotifierDecorator(notifier: facebookNotifierDecorator);

  slackNotifierDecorator.sendNotification();

  /// Here is the result:
  /*
  Sending SMS notification!
  Sending Facebook notification!
  Sending Slack notification!
  */
}

File: Swift/Structural/decorator.swift
//
//  decorator.swift
//
//
//

// Base protocol
protocol Notifier {
    func sendNotification()
}

// Concrete implementation for SMS notifications
struct SmsNotifier: Notifier {
    func sendNotification() { print("Sending SMS notification!") }
}

// Concrete implementation for Facebook notifications
struct FacebookNotifier: Notifier {
    func sendNotification() { print("Sending Facebook notification!") }
}

// Concrete implementation for Slack notifications
struct SlackNotifier: Notifier {
    func sendNotification() { print("Sending Slack notification!") }
}

// Base decorator class
// Wraps any Notifier and delegates to it
class NotifierDecorator: Notifier {
    private let wrappee: any Notifier

    init(notifier: any Notifier) {
        self.wrappee = notifier
    }

    func sendNotification() {
        wrappee.sendNotification()
    }
}

// Extension for SMS notifications
class SmsNotifierDecorator: NotifierDecorator {
    override func sendNotification() {
        super.sendNotification()
        print("Sending SMS notification!")
    }
}

// Extension for Facebook notifications
class FacebookNotifierDecorator: NotifierDecorator {
    override func sendNotification() {
        super.sendNotification()
        print("Sending Facebook notification!")
    }
}

// Extension for Slack notifications
class SlackNotifierDecorator: NotifierDecorator {
    override func sendNotification() {
        super.sendNotification()
        print("Sending Slack notification!")
    }
}

// Demo
let smsNotifier = SmsNotifier()
let facebookNotifierDecorator = FacebookNotifierDecorator(notifier: smsNotifier)
let slackNotifierDecorator = SlackNotifierDecorator(notifier: facebookNotifierDecorator)

slackNotifierDecorator.sendNotification()

// Here is the result:
// Sending SMS notification!
// Sending Facebook notification!
// Sending Slack notification!
